src/include/simeng/config/SimInfo.hh

```diff
@@ -0,0 +1,259 @@
+#pragma once
+
+#include <iostream>
+#include <string>
+
+#include "simeng/Instruction.hh"
+#include "simeng/RegisterFileSet.hh"
+#include "simeng/config/ModelConfig.hh"
+#include "simeng/config/yaml/ryml.hh"
+
+namespace simeng {
+namespace config {
+
+#define DEFAULT_STR "Default"
+
+/** Enum representing the possible simulation modes. */
+enum class simMode { emulation, inorder, outoforder };
+
+/** A SimInfo class to hold values, specified by the constructed ryml::Tree
+ * object in the ModelConfig class and manually, used after the instantiation of
+ * the current simulation and its objects. */
+class SimInfo {
+ public:
+  /** A getter function to retrieve the ryml::Tree representing the underlying
+   * model config file. */
+  static ryml::ConstNodeRef getConfig() {
+    return getInstance()->validatedConfig_.crootref();
+  }
+
+  /** A setter function to set the model config file from a path to a YAML file.
+   */
+  static void setConfig(std::string path) { getInstance()->makeConfig(path); }
+
+  /** A function to add additional config values to the model config file. */
+  static void addToConfig(std::string configAdditions) {
+    getInstance()->mdlCnf_.addConfigOptions(configAdditions);
+    // Replace the validated config with new instance with the supplied
+    // additional values
+    getInstance()->validatedConfig_ = getInstance()->mdlCnf_.getConfig();
+    // Update previously extracted values from the config file
+    getInstance()->extractValues();
+  }
+
+  /** A function to generate a default config file based on a passed ISA. */
+  static void generateDefault(ISA isa, bool force = false) {
+    if (isa == config::ISA::AArch64)
+      getInstance()->mdlCnf_.reGenerateDefault(ISA::AArch64, force);
+    else if (isa == config::ISA::RV64)
+      getInstance()->mdlCnf_.reGenerateDefault(ISA::RV64, force);
+
+    // Update config path to be the default string
+    getInstance()->setConfigPath(DEFAULT_STR);
+
+    // Replace the validated config with the new default config
+    getInstance()->validatedConfig_ = getInstance()->mdlCnf_.getConfig();
+    // Update previously extracted values from the config file
+    getInstance()->extractValues();
+  }
+
+  /** A utility function to get a value, of a specified type, from a config
+   * option. */
+  template <typename T>
+  static T getValue(ryml::ConstNodeRef node) {
+    T val;
+    node >> val;
+    return val;
+  }
```

Comment:
I may be missing somethig here, but given we store `validatedConfig` in SimInfo anyway, could we not have the argument of `getValue` be some string or struct of strings rather than require all classes that want to access a config value to either store a copy of the config or be passed one as an argument?

For example, have `config::SimInfo::getValue<bool>("LSQ-L1-Interface/Exclusive")` or `config::SimInfo::getValue<bool>({"LSQ-L1-Interface", "Exclusive"})`?

This would reduce the number of copies of the config tree needing to be passed around the project, and allow all classes to automatically have access to the config file via `SimInfo.hh`.

Reply:
This could also help with another issue [raised previously](https://github.com/UoB-HPC/SimEng/pull/307#discussion_r1200792935); as forward declaring this single function along with `class SimInfo` should remove any circular dependancy from an implementation using `archInfo.hh` files in each ISAs `arch` directory

Reply:
I had thought of this prior but hit many dead ends. Given that you can have any depth regarding the YAML hierarchy, I'm unsure of a good way to do this. 

The string-only param has the problem of excluding certain characters from YAML config options.

We could do a vector of strings instead but it doesn't seem user-friendly. However, there's an argument that the current solution also has an issue of "user-friendlyness".

Reply:
I guess it's a question of 
```
config::SimInfo::getValue<int>(config["Pipeline-Widths"]["FrontEnd"]
```
vs 
```
config::SimInfo::getValue<int>({"Pipeline-Widths", "FrontEnd"})
```

Reply:
One small other point would be that without the current method, you'd have to convert non-string values to strings before passing them as part of the param. E.g. having to convert the iteration count of a for loop from an int to a string.

Possible a minor detail but all this is for the sake of readability

Reply:
I would argue that a vector of strings is no less user friendly than the current setup - you still need to know the exact names of the fields anyway; you're just encasing it in a vector rather than in `[ ]`. Plus, in my opinion, not having to pass `config` to every class that would need information from the config file is a big plus, and also eliminates n copies of the same object being passed, created, or stored in the project as currently is the case.

I'm also not 100% sure on your final point about converting things to strings... wouldn't this be the case currently too? As field names are passed always as `["field name here"]`. Please correct me if this isn't the case though!

Reply:
I agree with the moot user-friendly point hence my previous comment
> However, there's an argument that the current solution also has an issue of "user-friendlyness".

Also, we're passing the config tree around as a const reference so not sure it has the detrimental impact of n actual copies.

Further to the n copies point, if you wanted to truly eliminate all copies then you'd have to replace all accesses to the config which don't use the `config::SimInfo::getValue` utility which were removed in the previous PR due to readability. For example,

```
 ryml::ConstNodeRef config = config::SimInfo::getConfig();
 config["Branch-Predictor"]["BTB-Tag-Bits"] >> btbBits_;
 config["Branch-Predictor"]["Saturating-Count-Bits"] >> satCntBits_;
 config["Branch-Predictor"]["Global-History-Length"] >> globalHistoryLength_;
 config["Branch-Predictor"]["RAS-entries"] >> rasSize_;
```
would become,
```
btbBits_ = config::SimInfo::getValue<uint64_t>({"Branch-Predictor", "BTB-Tag-Bits"});
satCntBits_ = config::SimInfo::getValue<uint8_t>({"Branch-Predictor", "Saturating-Count-Bits"});
globalHistoryLength_ = config::SimInfo::getValue<uint16_t>({"Branch-Predictor", "Global-History-Length"});
rasSize_ = config::SimInfo::getValue<uint64_t>({"Branch-Predictor", "RAS-entries"});
```
which gets worse when you consider each access has to be defined from the root of the YAML tree (e.g. you can't define the point of access as `ryml::ConstNodeRef port_node = config["Latencies"][i];` and then extract from there).

For the last point. Let's say you're doing something in a loop with iteration variable `i`, then we'd have:
```
config::SimInfo::getValue<bool>(config["Execution-Units"][i]["Pipelined"])
```
vs
```
config::SimInfo::getValue<bool>({"Execution-Units", std::to_string(i), "Pipelined"})
```

Reply:
Just realised in my full review that we pass around as reference, so yup I agree that is moot.

Given what you've said, I agree keeping it as is would be worthwhile, however, I do think we should use `SimInfo::getValue<uint64_t>` etc as much as possible mainly as I think `var = SimInfo::getValue<uint64_t>()` is more intuative than `config[...][...] >> var`